plugins/hls-pragmas-plugin/src/Ide/Plugin/Pragmas.hs

```diff
@@ -187,7 +188,19 @@ completion _ide _ complParams = do
 -- Useful for inserting pragmas.
 endOfModuleHeader :: T.Text -> Range
 endOfModuleHeader contents = Range loc loc
-    where
-        loc = Position line 0
-        line = maybe 0 succ (lastLineWithPrefix "{-#" <|> lastLineWithPrefix "#!")
-        lastLineWithPrefix pre = listToMaybe $ reverse $ findIndices (T.isPrefixOf pre) $ T.lines contents
+  where
+    loc = Position line 0
+    line = maybe 0 succ (lastShebang contents <|> lastOptsGhc contents <|> lastLangPrag contents)
+    lastShebang = lastLineWithPrefix $ T.isPrefixOf "#!"
+    lastLangPrag = lastLineWithPrefix $ checkPragma "LANGUAGE" 8
+    lastOptsGhc = lastLineWithPrefix $ checkPragma "OPTIONS_GHC" 11
+
+lastLineWithPrefix :: (T.Text -> Bool) -> T.Text -> Maybe Int
+lastLineWithPrefix p contents = listToMaybe $ reverse $ findIndices p $ T.lines contents
+
+checkPragma :: T.Text -> Int -> T.Text -> Bool
+checkPragma name n = check
+  where
+    check l = isPragma l && getName l == name
+    getName l = T.take n $ T.dropWhile isSpace $ T.drop 3 l
```

Comment:
It looks like the second argument is always the length of the first. Why not derive it internally if that's the case?

Reply:
hi, thanks for the quick feedback! 

Ah yes good point, I'll change that function now. I also realised after that there are a few cases where this won't work as expected - for example, it will add a new Language pragma before the other language pragmas if the there is a shebang at the start of the file -  like this:

#! /usr/bin/env nix-shell
{-# LANGUAGE OverloadedStrings #-} -- newly added, should be last
{-# LANGUAGE NamedFieldPuns #-}

so I was thinking it would be better to calculate the final line number as a sum the shebang, options_ghc and language pragmas. I did this locally and he works for all the older cases and new ones I could think of, like the one above. I can update my pr with this change this evening and also with the expanded comment? 

Reply:
I was wondering this too - which is why I asked for more comments. Maybe it would be a good idea to add a few more tests to exercise a few more scenarios ?

Reply:
I was wondering this too - which is why I asked for more comments. Maybe it would be a good idea to add a few more tests to exercise a few more scenarios ?

Reply:
Sure no problem, I can include a few more with the update.

Reply:
So I added a number of new tests, different combinations of shebangs, ghc_options, language pragmas and later inline pragmas. I can't really think of any other combinations right now, but do let me know if I missed something or if I could improve the solution. Thanks!

Reply:
hi again, I just cleaned up the code a bit, added another test for when 'module ...' is on the first line, and also squashed it down to one commit.